### ISSUE TITLE
[unix] Waiting for `Child` after `kill`, deadlocks

### ISSUE THREAD

**Issue by aslpavel** [alexcrichton/tokio-process#57]
Once `kill` is called `Child` future never resolves.

### Minimal code example
This program, will never terminate
```rust
use failure::Error;
use futures::Future;
use std::process::Command;
use tokio::runtime::Runtime;
use tokio_process::CommandExt;

fn run() -> impl Future<Item = (), Error = Error> {
    let mut child = Command::new("sleep").arg("1").spawn_async().unwrap();
    // if you comment next line everything works, otherwise process status will never be reported
    child.kill().unwrap();
    child
        .from_err()
        .map(|status| println!("process status: {:?}", status))
}

fn main() -> Result<(), Error> {
    let mut runtime = Runtime::new()?;
    runtime.block_on(run())?;
    Ok(())
}
```
It is reproducible on both Linux and MacOS

**Comment by ipetkov** [comment]
Hi @aslpavel thanks for the report!

I tried this example on my macOS machine and it resolved as expected even after multiple runs:

```
> cargo run --example issue_57
   Compiling tokio-process v0.2.3 (/Users/ivan/Desktop/tokio/tokio-process)
    Finished dev [unoptimized + debuginfo] target(s) in 4.02s
     Running `target/debug/examples/issue_57`
process status: ExitStatus(ExitStatus(9))
```

I wonder if this is a duplicate issue of #51. Could you verify if the deadlock occurs during the actual `kill` invocation, or during the call to `runtime.block_on`?

**Comment by aslpavel** [comment]
I've done some more debugging:
- public crate "0.2.3", definitely deadlocks inside `poll`
- if I build from master it works fine (so whatever was the problem it is fixed now), but it would be great if we could understand why it locks in the first place. My guess is that somewhere between `try_wait` and actually starting listening for `signals`, we are receiving `SIGCHLD`, and that is why `poll` hangs? Have any related bugs has been fixed recently?

Here is updated version with tracing
```rust
use failure::Error;
use futures::Future;
use std::process::Command;
use tokio::runtime::Runtime;
use tokio_process::CommandExt;

fn run() -> impl Future<Item = (), Error = Error> {
    let mut child = Command::new("sleep").arg("1").spawn_async().unwrap();
    // if you comment next line everything works, otherwise process status will never be reported
    println!("sending kill");
    child.kill().unwrap();
    println!("kill send");
    child
        .trace("wait child")
        .from_err()
        .map(|status| println!("process status: {:?}", status))
}

fn main() -> Result<(), Error> {
    let mut runtime = Runtime::new()?;
    runtime.block_on(run())?;
    Ok(())
}

// -----------------------------------------------------------------------------
// Trace
// -----------------------------------------------------------------------------
pub trait FutureExt: Future + Sized {
    fn trace<M: AsRef<str>>(self, msg: M) -> Trace<Self> {
        Trace {
            inner: self,
            message: msg.as_ref().to_owned(),
        }
    }
}

impl<F: Future> FutureExt for F {}

pub struct Trace<F> {
    inner: F,
    message: String,
}

impl<F: Future> Future for Trace<F> {
    type Item = F::Item;
    type Error = F::Error;

    fn poll(&mut self) -> futures::Poll<Self::Item, Self::Error> {
        println!("[TRACE][{}] poll begin", self.message);
        match self.inner.poll() {
            done @ Ok(futures::Async::Ready(_)) => {
                println!("[TRACE][{}] done", self.message);
                done
            }
            not_ready @ Ok(futures::Async::NotReady) => {
                println!("[TRACE][{}] not ready", self.message);
                not_ready
            }
            error @ Err(_) => {
                println!("[TRACE][{}] error", self.message);
                error
            }
        }
    }
}
```
- output for `tokio-process = "0.2.3"`, this is 100% reproducible for me
```bash
$ cargo run
    Updating crates.io index
    Finished dev [unoptimized + debuginfo] target(s) in 0.94s
     Running `target/debug/test-proc`
sending kill
kill send
[TRACE][wait child] poll begin
[TRACE][wait child] not ready
...
Never termiates
```

- build from master `tokio-process = { git = "https://github.com/alexcrichton/tokio-process" }`
```bash
$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.25s
     Running `target/debug/test-proc`
sending kill
kill send
[TRACE][wait child] poll begin
[TRACE][wait child] done
process status: ExitStatus(ExitStatus(9))
```

**Comment by aslpavel** [comment]
Ok, I know where bug was, looking at the code for `tag 0.2.3`. `Child` future violates contract of `Future` interface
- `kill`, calls `try_wait` which sets `reaped` to `true`
- [`Child::poll`](https://github.com/alexcrichton/tokio-process/blob/b81a627eb6125d7e7c1d30de4cc3bd958c772427/src/unix.rs#L104) returns `NotReady`, and since it has not registered to be notified, `poll` will never be called again

**Comment by aslpavel** [comment]
One more thing this is actually supposed to be common pattern, sending `kill` will leave zombie process behind, better semantic for publicly exposed `kill` should return `Future` which will `wait` for process to terminate. At the very least we should probably add this as part of documentation. Same goes for `drop` but it is less clear how to fix it, essentially current implementation of `drop` will always leaves zombie behind.

**Comment by ipetkov** [comment]
Ah yes, thank you for digging into this further!

Yes, the current master as some partial fixes for this:
* We no longer reap the child within kill (I believe this was done earlier to handle zombie processes if the `Child` handle was dropped, but given that we publicly expose the same method, reaping there probably isn't the right thing)
* The "is reaped" guard that is causing the deadlock has also been optimized out (the underlying `std::process::Child` basically implements this already, so now we lean on it [which will keep returning the same status if it is "polled" again])

As far as blocking in drop, it seems like something that should be avoided, given the reports in #51. I was imagining a solution where any orphaned zombie processes are stuck in some global queue where we can eventually try reaping them again (but avoid blocking any particular drop). This part hasn't been done yet

**Comment by ipetkov** [comment]
Given that the immediate cause of the deadlocks has been fixed in master, and given that we're already tracking a broader fix around this in #51, let's move the discussion there.

Thanks again for the deep dive @aslpavel!
